doc/website/blog/2022-03-14-ent-hooks-examples.md

```diff
@@ -0,0 +1,434 @@
+
+In this post, I will focus mainly on schemas hooks.
+
+### So why use them?
+
+In ent, we try and push the entity's logic down to the schema level, this allows us to define everything about an entity close to its creation. Entity properties could be:
+1. Valid values.
+2. Types.
+3. Allowed mutations.
+4. Privacy.
+5. Required Side effect (heavy compute for example).
+
+---
+Many times, it’s more comfortable to define that in the RPC level, a controller provides a context and all the mutations
+and validation are done before changing the value.
+Yet, as the system gets bigger and more developers are working on it making sure that these validations and calculations are done
+whenever the entity is changing becomes harder and requires more pre-known knowledge about each entity to work with and
+can cause unintentional bugs or inconsistencies in the data that are only discovered in runtime.
+The way ent, based on the experience of bigger companies like Facebook, decided to handle it is to keep this information
+as close as possible to the entity, therefore hooks are just that, a piece of code that runs before and after an entity 
+is changing and defines that additional behavior.
+
+## Let us do some examples
+
+For this examples we use a simple schema of users and dogs.
+Each user can have multiple pets and each pet has an owner.
+We have also added a cache entity that will be used later.
+as can be seen here:
+```go
+// User holds the schema definition for the User entity.
+type User struct {
+	ent.Schema
+}
+
+// Fields of the User.
+func (User) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("name").
+			NotEmpty(),
+		field.String("connection_string").
+			NotEmpty(),
+		field.String("password").
+			Sensitive().
+			NotEmpty(),
+	}
+}
+
+// Edges of the User.
+func (User) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("pets", Dog.Type),
+		edge.To("cache", Cache.Type).Unique(),
+	}
+}
+```
+```go
+// Dog holds the schema definition for the Dog entity.
+type Dog struct {
+	ent.Schema
+}
+
+// Fields of the Dog.
+func (Dog) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("name").
+			NotEmpty(),
+		field.Int("owner_id").
+			Optional(),
+	}
+}
+
+// Edges of the Dog.
+func (Dog) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("owner", User.Type).
+			Ref("pets").
+			Field("owner_id").
+			Unique(),
+	}
+}
+```
+```go
+// Cache holds the schema definition for the Cache entity.
+type Cache struct {
+	ent.Schema
+}
+
+// Fields of the Cache.
+func (Cache) Fields() []ent.Field {
+	return []ent.Field{
+		field.Int("walks"),
+	}
+}
+
+
+```
+
+During each example I'll provide a test and a recipe, I really suggest starting with a test to see that the hook does what you need it to do,
+Unlike many other parts of the codebase, hooks do one thing only and should be simple to test.
+
+### Keeping our secrets safe:
+
+Many times, we need to store strings that include secrets, a good example can be a connection
```

Comment:
We might wanna use something different here. In a `User`, ` Dog` story a connection string is a weird property :-D 

Maybe use your chip number and hide that or whatever :laughing:

Reply:
as @masseelch wrote, this is a neat example of using a hook but if you're trying to build a thin narrative around the examples you're giving, do the extra work of making your story coherent. 

btw, I think you can live with just a few standalone examples, but since you presented this like there is some narrative, this example feels weird